src/app/clusters/device-energy-management-server/device-energy-management-server.cpp

```diff
@@ -226,19 +236,88 @@ void Instance::InvokeCommand(HandlerContext & handlerContext)
                 [this](HandlerContext & ctx, const auto & commandData) { HandleRequestConstraintBasedForecast(ctx, commandData); });
         }
         return;
+    case CancelRequest::Id:
+        if (!HasFeature(Feature::kStartTimeAdjustment) && !HasFeature(Feature::kForecastAdjustment) &&
+            !HasFeature(Feature::kConstraintBasedAdjustment))
+        {
+            handlerContext.mCommandHandler.AddStatus(handlerContext.mRequestPath, Status::UnsupportedCommand);
+        }
+        else
+        {
+            HandleCommand<CancelRequest::DecodableType>(
+                handlerContext, [this](HandlerContext & ctx, const auto & commandData) { HandleCancelRequest(ctx, commandData); });
+        }
+        return;
     }
 }
 
-void Instance::HandlePowerAdjustRequest(HandlerContext & ctx, const Commands::PowerAdjustRequest::DecodableType & commandData)
+Status Instance::CheckOptOutAllowsRequest(AdjustmentCauseEnum adjustmentCause)
 {
-    int64_t power        = commandData.power;
-    uint32_t durationSec = commandData.duration;
-    bool validArgs       = false;
-    Status status        = Status::Success;
+    OptOutStateEnum optOutState = mDelegate.GetOptOutState();
+
+    if (adjustmentCause >= AdjustmentCauseEnum::kUnknownEnumValue)
+    {
+        ChipLogError(Zcl, "DEM: adjustment cause is invalid (%d)", static_cast<int>(adjustmentCause));
+        return Status::InvalidValue;
+    }
+
+    switch (optOutState)
+    {
+    case OptOutStateEnum::kNoOptOut: /* User has NOT opted out so allow it */
+        ChipLogProgress(Zcl, "DEM: OptOutState = kNoOptOut");
+        return Status::Success;
+
+    case OptOutStateEnum::kLocalOptOut: /* User has opted out from Local only*/
```

Comment:
Nothing in the spec has normative requirements that clearly correspond to this method.  That needs to get fixed in the spec.  Please make sure a spec issue is filed.

To be precise, the spec says:

> the ESA SHALL validate that ... the OptOutState permits the specified AdjustmentCauseEnum

but nothing clearly defines in a normative way what adjustment causes are allowed in which OptOutState values.  The "Summary" column for enums is not normative.

Reply:
https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/8766 raised